### PR TITLE
[ADAM-1164] Add parallel file merger.

### LINKED ISSUE #1164: Merging files should be multithreaded
Related to #1161. Right now, we merge files as a big ol' single threaded hunk of code. However, this can be parallelized.

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #1164. I'd actually love to get this into 0.22.0 as well. Thoughts?

### CODE REVIEW

**Comment by fnothaft** [comment]
Just following up with runtime numbers. With this, saving the NA12878 234GB BAM back to a single BAM from Parquet runs in 4.4 minutes on 833 cores (2.4 minutes to go ADAM->BAM, 2.0 minutes to do the merge). Without this, it takes 44 minutes (2.3 minutes to go ADAM->BAM, the remainder to merge).

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ParallelFileMerger.scala [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    val blockSize = fs.getDefaultBlockSize(outputPath)
+
+    // generate the merge list
+    val merges = generateMerges(blockSize, sizeByFile)
+
+    // make a temp output path
+    //
+    // ideally, this would be a directory, however, fs.concat has the
+    // undocumented contract that the paths being merged must live in
+    // the same directory as the path they are being merged to
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ParallelFileMerger.scala]: what? hope you didn't have to find that out the hard way

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ParallelFileMerger.scala [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        broadcastConf,
+        gzipAtEnd,
+        cramAtEnd,
+        bufferSize)
+    })
+
+    // UNDOCUMENTED in hadoop fs API:
+    // all paths passed to the concat method must be qualified with
+    // full scheme and name node URI
+    val outputPaths = (0 until numBlocksToWrite).map(idx => {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ParallelFileMerger.scala]: ...and sigh

**Comment by heuermh** [comment]
Thank you, @fnothaft!

**Pull request merged**